### ISSUE TITLE
Clickhouse uses 3 * partition_size RAM to move partition to S3

### ISSUE THREAD

**Issue by vmihailenco** [ClickHouse/ClickHouse#20179]
I try to move a partition with a single part which size is reported to be 1.40 GiB (by system.parts):

```sql
alter table tbl_name move partition 'partition' to volume 's3'
```

And CH says it can't allocate 5.83 GiB (server indeed does not have that much memory):

```
2021.02.07 11:48:01.741520 [ 33327 ] {4bfd5fc4-3c29-49fe-9106-98a6e08dff97} <Error> TCPHandler: Code: 241, e.displayText() = DB::Exception: Memory limit (total) exceeded: would use 5.83 GiB (attempt to allocate chunk of 3221225472 bytes), maximum: 4.54 GiB, Stack trace:

0. DB::Exception::Exception<char const*, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*&&, char const*&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&) @ 0x7d8e410 in /usr/bin/clickhouse
1. MemoryTracker::alloc(long) @ 0x7d8df41 in /usr/bin/clickhouse
2. operator new(unsigned long) @ 0x7d8b7d8 in /usr/bin/clickhouse
3. std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::append(unsigned long, char) @ 0x123f15c0 in ?
4. DB::WriteBufferFromVector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::nextImpl() @ 0x7d9de6a in /usr/bin/clickhouse
5. DB::WriteBufferFromS3::nextImpl() @ 0xd9bf887 in /usr/bin/clickhouse
6. ? @ 0xd9bc81b in /usr/bin/clickhouse
7. DB::WriteBuffer::write(char const*, unsigned long) @ 0x7d9e16e in /usr/bin/clickhouse
8. DB::copyFile(DB::IDisk&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::IDisk&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xd99b26b in /usr/bin/clickhouse
9. ? @ 0xd99c908 in /usr/bin/clickhouse
10. DB::AsyncExecutor::execute(std::__1::function<void ()>)::'lambda'()::operator()() const @ 0xd9b8bc7 in /usr/bin/clickhouse
11. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x7dc5b3d in /usr/bin/clickhouse
12. ? @ 0x7dc96f3 in /usr/bin/clickhouse
13. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
14. clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
```

Is it possible to improve CH to not allocate 5.83 GiB in order to move 1.40 GiB part? Perhaps tweak some settings? ClickHouse 20.12.5.18

**Comment by alexey-milovidov** [comment]
S3 volume is experimental feature in development stage, it has obvious drawbacks that should be eliminated before production readiness. They use O(N) memory for files - it is just unbearable, and before it is fixed, you should not consider the feature for any serious usage.

**Comment by vmihailenco** [comment]
Thanks for the answer. I guess I will continue to use it with caution since otherwise it is working well enough and S3 pricing is considerably lower (and zero maintenance).

If there are any workarounds (like ability to split parts) - please let me know.

**Comment by alex-zaitsev** [comment]
@excitoon will look into that

**Comment by Jokser** [comment]
Probably this fix should help with memory consumption when DiskS3 is used:
https://github.com/ClickHouse/ClickHouse/pull/17934

**Comment by vmihailenco** [comment]
@Jokser yes, it sounds exactly like the problem I have. Not sure if it was backported or not - PR is closed... But I definitely don't have that setting in 20.12.5 .

Not sure if it is a problem or not but it is strange that `s3_max_single_part_upload_size < s3_min_upload_part_size`:

```
│ s3_min_upload_part_size        │ 536870912 │       0 │ The minimum size of part to upload during multipart upload to S3.   │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ UInt64 │
│ s3_max_single_part_upload_size │ 67108864  │       0 │ The maximum size of object to upload using singlepart upload to S3. │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ UInt64 │
```

**Comment by Jokser** [comment]
@vmihailenco
For 20.12 you can try to set the S3 disk option `min_multi_part_upload_size` to `0` to enforce multipart upload mode for all writes to S3. It should help with extra memory allocations in WriteBufferFromS3.
How It should look like in configuration:

```
<disks>
 <s3>
   <endpoint>...</endpoint>
   <min_multi_part_upload_size>0</min_multi_part_upload_size>
 </s3>
</disks>
```

It's not strange that `s3_max_single_part_upload_size < s3_min_upload_part_size`.
If amount of data in buffer is less than `s3_max_single_part_upload_size` then singlepart upload will be used.
If greater data writes in WriteBufferFromS3 will be batched by `s3_min_upload_part_size` size.

**Comment by vmihailenco** [comment]
@Jokser thanks, I've tried with min_multi_part_upload_size=0 and min_multi_part_upload_size=1 without much success (same behavior and error). Do I need to manually restart clickhouse-server? It is not clear since I don't have such setting in `system.settings` and I don't know what to check else.

I was not able to test this *after* the restart due to a different problem. And now I am out of my incidents-per-week budget :)

**Comment by Jokser** [comment]
@vmihailenco
Yes, you need to restart clickhouse-server to apply this option.
Setting `min_multi_part_upload_size=1` doesn't make sense, only `min_multi_part_upload_size=0` may help with memory consumption.

**Comment by vmihailenco** [comment]
Does anyone know what memory usage should I expect from multi-part upload? I would expect CH to use 100-200MB buffer but it does not seem to be the case. Does it try to merge parts before uploading or something?

It looks like `min_multi_part_upload_size=0` helps but memory usage is still several gigabytes...

**Comment by pumelo** [comment]
This one caught me too. It was not pointed out but let me add: **This leads to a corrupted database and lost data!**
If using `alter table xxx.yyyy move partition '2020-10' to volume 's3'` the files uploaded to s3 are incomplete (due to memory error during upload), but the part is removed from local disk.

As @alex-zaitsev pointed out, don't use s3, it's not ready yet. 😢

**Comment by Jokser** [comment]
@pumelo @vmihailenco 
Decreasing these options
s3_min_upload_part_size
s3_max_single_part_upload_size
to 32Mb should help with memory consumption if you use CH version 21.1

They should be placed under 'default' profile section in users.xml config file.

**Comment by alexey-milovidov** [comment]
DiskS3 is experimental and it won't be considered production ready until it will use O(1) memory.
I have no idea why someone is advertising this feature.

**Comment by Jokser** [comment]
@alexey-milovidov 
Since 21.1 (after this fix https://github.com/ClickHouse/ClickHouse/pull/17934) memory consumption for S3 is O(1) and it's equal to max (s3_min_upload_part_size, s3_max_single_part_upload_size) for each part column file.

**Comment by vmihailenco** [comment]
I can confirm that it is working in 21.2 even with default settings.

>for each part column file.

@Jokser Do you mean that each column in a part is uploaded in parallel? That could explain why memory usage is still larger than default 512mb (s3_min_upload_part_size=536870912). But it also means that memory usage varies a lot depending on number of columns in a table...

**Comment by Jokser** [comment]
@vmihailenco Yes, each column is uploaded in parallel. But since 21.3 version you can limit the number of threads used in parallel copies and other async operations for S3 disk: `thread_pool_size` which defaults to 16.

**Comment by vmihailenco** [comment]
Thanks. CH is always a step ahead :)

Ideally thread_pool_size should not be larger than number of cpu (or 2 * num_cpu), but I guess I can tweak this myself.

Closing since this seems to be fixed in 21.3 with combination of:
- s3_min_upload_part_size
- thread_pool_size
